fili-core/src/main/java/com/yahoo/bard/webservice/web/FilterOptimizingRequestMapper.java

```diff
@@ -0,0 +1,61 @@
+// Copyright 2019 Verizon Media Group.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.web;
+
+import com.yahoo.bard.webservice.data.FilterOptimizable;
+import com.yahoo.bard.webservice.data.config.ResourceDictionaries;
+import com.yahoo.bard.webservice.data.dimension.Dimension;
+import com.yahoo.bard.webservice.web.apirequest.DataApiRequest;
+import com.yahoo.bard.webservice.web.filters.ApiFilters;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import javax.ws.rs.container.ContainerRequestContext;
+
+/**
+ * Request mapper that checks if any of the Dimensions that are being filtered on can optimize their filters, and if
+ * so performs that optimization.
+ */
+public class FilterOptimizingRequestMapper extends RequestMapper<DataApiRequest> {
+
+    /**
+     * Constructor.
+     *
+     * @param resourceDictionaries The dictionaries to use for request mapping.
+     */
+    public FilterOptimizingRequestMapper(
+            ResourceDictionaries resourceDictionaries
+    ) {
+        super(resourceDictionaries);
+    }
+
+    @Override
+    public DataApiRequest apply(DataApiRequest request, ContainerRequestContext context)
+            throws RequestValidationException {
+        if (request.getApiFilters() == null || request.getApiFilters().isEmpty()) {
+            return request;
+        }
+
+        ApiFilters newFilters = new ApiFilters();
+        for (Map.Entry<Dimension, Set<ApiFilter>> entry : request.getApiFilters().entrySet()) {
+            if (! (entry.getKey() instanceof FilterOptimizable)) {
+                newFilters.put(entry.getKey(), entry.getValue());
+                continue;
+            }
```

Comment:
This is a bug.
If Dimension A is processed and optimizes filters into Dimension B, and then Dimension B is processed, this put will cause it's filters to overwrite the filters that were originally dimension A.

Probably the best approach here is to create a List of entries and push values into it and then do a grouping collect at the end.